sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp

```diff
@@ -290,11 +308,57 @@ class wi_element<sycl::ext::oneapi::bfloat16, NumRows, NumCols, Use, Layout,
 #endif // __SYCL_DEVICE_ONLY__
 };
 
-} // namespace matrix
-} // namespace experimental
-} // namespace oneapi
+// End wi_element definition
+
+// Begin wi_data definition
+
+template <typename Group, typename T,
+          sycl::ext::oneapi::experimental::matrix::use Use, size_t Rows,
+          size_t Cols, sycl::ext::oneapi::experimental::matrix::layout Layout>
+class wi_data {
+
+  sycl::ext::oneapi::experimental::matrix::joint_matrix<Group, T, Use, Rows,
+                                                        Cols, Layout> &jm;
+
+  wi_data(sycl::ext::oneapi::experimental::matrix::joint_matrix<
+          Group, T, Use, Rows, Cols, Layout> &_jm)
+      : jm(_jm){};
+
+  template <typename Grp, typename Type,
+            sycl::ext::oneapi::experimental::matrix::use UseJm, size_t NumRows,
+            size_t NumCols,
+            sycl::ext::oneapi::experimental::matrix::layout LayoutJm>
+  friend decltype(auto)
+  get_wi_data(Grp, sycl::ext::oneapi::experimental::matrix::joint_matrix<
+                       Grp, Type, UseJm, NumRows, NumCols, LayoutJm> &);
+
+public:
+  size_t length() {
+#if __SYCL_DEVICE_ONLY__
+    return __spirv_JointMatrixWorkItemLengthINTEL(jm.spvm);
+#else
+    throw runtime_error("joint matrix is not supported on host device.",
+                        PI_ERROR_INVALID_DEVICE);
+#endif
+  };
+
+  decltype(auto) operator[](size_t i) {
+    return wi_element<T, Rows, Cols, Use, Layout, Group>(jm, i);
```

Comment:
So you've moved `wi_data` to the intel header but it is still in the oneapi namespace. I'm not sure it is a good idea to immediately remove the nvidia implementation from wi_data immediately. There is code that is already using this extension and it could be wiser to keep the implementation for now, whilst deprecating it and adding the nvidia implementation to `joint_matrix_apply` at the same time.

Also, if you move wi_data to the intel namespace then you can either just still do the above (existing code use won't care since wi_data is returned as a auto placeholder anyway), and you could even deprecate it in the oneapi namespace and add it to the intel name space also whilst only implementing it there for intel backends.

Reply:
@JackAKirk , Are you OK if I keep the NVIDIA implementation in the oneapi namespace (in `matrix-unified.hpp`) and the `jm` based implementation in the `intel` namespace in `matrix-intel.hpp`?  Thank you

Reply:
Yes I think so, although I'm not completely sure I understand what you mean.
Basically from the Nvidia implementation point of view I don't think it matters so long as the changes don't immediately break code that is using `get_wi_data()`. Then wherever the Nvidia implementation is moved to it will be soon removed anyway, once we have given people time to switch to using the `joint_matrix_apply`.

Thanks

Reply:
Basically so long as changes don't break this test: https://github.com/intel/llvm-test-suite/blob/intel/SYCL/Matrix/element_wise_all_ops_cuda.cpp

Reply:
@JackAKirk I can keep the NVIDIA/CUDA implementation (I hope they mean the same thing?) in the intel namespace.  In that case, we just have to modify the `get_wi_data()` call to use the `intel` namespace.  Are you OK with that code change in the test case?  I was leaning towards keeping the CUDA implementation in the `oneapi` namespace (in the m`matrix-unified.hpp`) so that we do not have to update the `get_wi_data()` call in the existing NVIDIA test cases.  Let me know what you think.

Reply:
Ah I see, yeah you are right I think that `get_wi_data()` for nvidia should stay in the oneapi namespace for Nvidia if that is OK. Point being that if it moved it would break existing code.

Yeah by Nvidia and CUDA backends I mean the same thing. Thanks